test/shared/src/main/scala/zio/test/Gen.scala

```diff
@@ -84,26 +190,88 @@ object Gen {
     Gen(ZStream.fromEffect(ZIO.accessM[Random](r => f(r.random))))
 
   /**
-   * A generator of integral values inside the specified range: [start, end).
+   * A generator of integers inside the specified range: [start, end].
+   * The shrinker will shrink toward the lower end of the range ("smallest").
+   */
+  final def int(min: Int, max: Int): Gen[Random, Int] =
+    integral(min, max)
+
+  /**
+   * A generator of integral values inside the specified range: [start, end].
+   * The shrinker will shrink toward the lower end of the range ("smallest").
+   */
+  final def integral[A](min: A, max: A)(implicit I: Integral[A]): Gen[Random, A] =
+    fromEffectSample {
+      nextInt(I.toInt(max) - I.toInt(min) + 1)
+        .map(r => I.plus(min, I.fromInt(r)))
+        .map(Sample.shrinkIntegral(min))
+    }
+
+  final def listOf[R, A](min: Int, max: Int)(g: Gen[R, A]): Gen[R with Random, List[A]] =
```

Comment:
I was thinking it could be more _compositional_ to make the "sized" variants like this:

```scala
final def listOf(g: Gen[R, A]): Int => Gen[R with Random, List[A]]`
```

Then you use them like this:

```scala
import Gen.{ listOf, sized }

sized(0, 10)(listOf(boolean))
```

What do you think?

Reply:
By the way, a new pattern that infers better is using `R <: Random` on the type declaration, rather than `R with Random` on the output.

Reply:
I like that.

It potentially takes us in the direction of being able to have test modes where the size parameter is controlled by the runner versus the test, which a lot of other frameworks have but we don't yet. You could imagine having an alias for `Int => Gen[R, A]`, maybe `SGen[R, A]` for a `Gen` that takes a size parameter. Then we could have a `checkSized` method that took an `SGen`, feed it a series of sizes, and checked each one.

The one issue I see here is that if you go down this road the next step is that you want to be able to compose sized generators, which works but would require writing a ton more code to make methods available on sized generators in addition to normal ones. One other way we could approach it is to use the environment type:

```scala
final case class Size(n: Int)

final def listOf[R <: Random with Size, A](g: Gen[R, A]): Gen[R, List[A]] =
  Gen.fromEffect(ZIO.access[Size](_.n)).flatMap(listOfN(_)(g))
```

Then sizes naturally compose together unless you want to provide a different size parameter locally.

Depending on where we go with this the `listOfN`, `StringN` and `vectorOfN` methods may be redundant with the new "sized" variants (they would be literally identical if the function versions if we just curried the parameters in the other order) so I think we take them out unless we go in a different direction here.

Reply:
I agree this makes sense. Although for namespacing reasons, it would probably be:

```scala
trait Size {
  def size: Int
}
```

Or maybe even better:

```scala
trait TestConfig {
  def testConfig: TestConfig.Service
}
```

...since we could stick other settings in there, as well.

I think we need to do this at some point, question is: 1.0 or after 1.0? 

Global size parameters were kinda abused in QuickCheck (et al). What size do you need for trees? What size do you need for 2D arrays? Is one global default really suitable for all cases?

Basically people use `sized` for convenience even when it didn't make sense, yielding some suboptimal tests. At least being explicit now, burden is on the user for choosing correct settings.

What I would really like to see in v1.0 and which should be easy is "small", "medium", and "large" combinators, e.g. `small(listOf(...))`, where the bigger ones spend more time on smaller sizes, but bump up to bigger ones once in a while (exponential or quadratic distribution).

Reply:
I agree with you about the dangers of relying too much on a global "size" parameter. My initial hypothesis would be that only collection types should have specific sized combinators (e.g. `listOf`) and in other cases users can always use the generic `sized` combinator.

On the naming, the downside of using something like `TestConfig`  is that it hides what the actual dependencies of a generator are. I really like how in the current API you can see right away whether a method generates random values by the `Random` environment parameter. I think it could be nice to do the same thing with `Size` so you can see that methods like `listOf` take a size parameter and other ones don't. Of course the downside is that if we end up having a lot more dependencies the type signatures get very verbose, but I'm not sure there are any other dependencies that generic test combinators should have to rely on.

My vote would be for getting this in now if we think it makes sense. The code base is still small and we can iterate quite quickly so might as well lay a strong foundation now so we are building in the right direction.

Shall I submit a PR on this so we can work through it?

Reply:
@adamgfraser All right, let's do it!

I'd propose:

```scala
trait SizedGen {
  def sizedGen: FiberRef[Int]
}
```

then we can augment the managed mock environment with `SizedGen` using some suitable default. 

The `sized` combinator now can simply require `R <: SizedGen`, and we can use that in the `list*` etc. methods. 

Sound reasonable?

Reply:
Yes!